### PR TITLE
Implement LoginServlet Test

### PR DESCRIPTION
**Pull request by ethyus**

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 1]
```diff
@@ -54,18 +57,33 @@ public void checkUserInDatastore(User user) {
...
+    try {
+      userEntity = datastore.get(userKey);
+      user.setUserKey(KeyFactory.keyToString(userKey));
+    } catch (EntityNotFoundException e) {
+      userEntity = null;
+    }
 
-        datastore.put(userEntity);
+    if (userEntity == null) {
+        addUserToDatastore(datastore, user);
```
Comment (ESaracay) [inline src/main/java/com/google/sps/servlets/LoginServlet.java:72]: If the userId is null then you should call the addUserToDatastore when you check on line 48. That way you don't also make a unnecessary request to datastore.

File: src/test/java/com/google/sps/LoginServletTest.java [hunk 2]
```diff
@@ -57,28 +72,60 @@ public void setUp() throws Exception {
...
   public void tearDown() throws Exception {
       helper.setEnvIsLoggedIn(true);
       helper.tearDown();
   }
+  
+  
+  @Test
+  public void userInDatastoreAlready() throws Exception {
+
+      datastore.put(USER_A.createEntity());
```
Comment (ESaracay) [inline src/test/java/com/google/sps/LoginServletTest.java:90]: Include a comment that explains what each test is checking. That way we can come back later and add edge cases without having to figure out what each test is doing.

**Review by Froggerson — APPROVED** [review]

File: src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 3]
```diff
@@ -29,22 +29,25 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+      userId = userService.getCurrentUser().getUserId();
+      logoutUrl = userService.createLogoutURL(urlToRedirect);
+      showNewTab = true;
     } else {
-      String loginUrl = userService.createLoginURL(urlToRedirect);
+      loginUrl = userService.createLoginURL(urlToRedirect);
     }
 
-    User user = new User(userId, loginUrl, logoutUrl, email, showNewTab);
+    User user = new User(userId, loginUrl, logoutUrl, userEmail, showNewTab);
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/LoginServlet.java:46]: Why are we storing the login and logout URLs, and whether to show a new tab in the user object and datastore?

These are ephemeral values and don't need to be stored.

File: src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 4]
```diff
@@ -29,22 +29,25 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
-      String logoutUrl = userService.createLogoutURL(urlToRedirect);
-      String showNewTab = true;
+      userEmail = userService.getCurrentUser().getEmail();
+      userId = userService.getCurrentUser().getUserId();
+      logoutUrl = userService.createLogoutURL(urlToRedirect);
+      showNewTab = true;
     } else {
-      String loginUrl = userService.createLoginURL(urlToRedirect);
+      loginUrl = userService.createLoginURL(urlToRedirect);
     }
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/LoginServlet.java:44]: Restructure this code to

```
User user;
if (userService.isUserLoggedIn()) {
  user = ...
} else {
  user = ...
}
```

This avoids the extra boolean.

File: src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 5]
```diff
@@ -54,18 +57,33 @@ public void checkUserInDatastore(User user) {
     // ** NEED TO WRITE TEST FOR BEHAVIORAL TESTING
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/LoginServlet.java:57]: This function 1) is not named properly and 2) does too many things.

1) The name of the function does not indicate to the programmer that it will store the user if they do not exist in the datastore.
2) A better way to write this logic would be to *not* add the user if they do not exist. Helper functions should generally do exactly one thing; this is currently doing two (checking and inserting).

So instead, you should make the function only check whether the user is in datastore, and then in your doGet, you could write

```
if (!isUserInDatastore(user)) {
  storeUserInDatastore(user);
}
```

File: src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 6]
```diff
@@ -54,18 +57,33 @@ public void checkUserInDatastore(User user) {
...
+      userEntity = null;
+    }
 
-        datastore.put(userEntity);
+    if (userEntity == null) {
+        addUserToDatastore(datastore, user);
     }
   }
+
+  public void addUserToDatastore(DatastoreService datastore, User user) {
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/LoginServlet.java:76]: `storeUserInDatastore`. In this context, store (or a similar word) is a less ambiguous word than add (adding a user to a database could mean creating a new user account with permissions to access the database).

File: src/test/java/com/google/sps/LoginServletTest.java (hunk 2 above)
Comment (tylerhou) [reply src/test/java/com/google/sps/LoginServletTest.java:90]: I think it's better to put that information into the name of the test function -- that way, when a test fails, the test runner will print the name of the function that failed.

In this case though, the function name could add additional information to help the reader understand the test. I would propose renaming the test to "loginWhenUserInDatastoreAlready".

(If the name of the function would be too long -- then write a comment.)

File: src/main/java/com/google/sps/data/Card.java [hunk 7]
```diff
@@ -104,7 +103,7 @@ public Entity createEntity(Key folderKey) {
         return card;
     }
```
Comment (tylerhou) [inline src/main/java/com/google/sps/data/Card.java:104]: How did this code get into the main branch? I don't see a pull request for it.

File: src/main/java/com/google/sps/data/Card.java [hunk 8]
```diff
@@ -104,7 +103,7 @@ public Entity createEntity(Key folderKey) {
...
 
-    public static Card EntityToCard(Entity entity) throws EntityNotFoundException {
+    public static Card EntityToCard(Entity entity) {
```
Comment (tylerhou) [inline src/main/java/com/google/sps/data/Card.java:106]: Maybe make this into constructor?

File: src/main/java/com/google/sps/data/Folder.java [hunk 9]
```diff
@@ -57,13 +52,13 @@ public Entity createEntity(Key userKey) {
     }
 
     // Returns a folder instance from a given Entity
-    public static Folder EntityToFolder(Entity entity) throws EntityNotFoundException {
+    public static Folder EntityToFolder(Entity entity) {
```
Comment (tylerhou) [inline src/main/java/com/google/sps/data/Folder.java:55]: Constructor?

File: src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 4 above)
Comment (tylerhou) [reply src/main/java/com/google/sps/servlets/LoginServlet.java:44]: Ping?

File: src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 5 above)
Comment (tylerhou) [reply src/main/java/com/google/sps/servlets/LoginServlet.java:57]: Ping?

File: src/main/java/com/google/sps/servlets/UserCardsServlet.java [hunk 10]
```diff
@@ -72,16 +75,14 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
 
-        String textNotTranslated = request.getParameter("textNotTranslated");
-
         // if blobKey is null, user did not add an image
         String folderKey = request.getParameter("folderKey");
         String blobKey = getBlobKeyfromBlobstore(request, "image");
         String labels = request.getParameter("labels");
         String fromLang = request.getParameter("fromLang");
         String toLang = request.getParameter("toLang");
         String textNotTranslated = request.getParameter("textNotTranslated");
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/UserCardsServlet.java:84]: `rawText`? For both the query parameter and the variable name.

File: src/main/java/com/google/sps/servlets/UserFoldersServlet.java [hunk 11]
```diff
@@ -34,15 +42,17 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
       String userKey = request.getParameter("userKey");
 
       // Query all folders identified by the userKey
-      Query folderQuery = new Query("Folder").setAncestor(userKey);
+      Query folderQuery = new Query("Folder").setAncestor(KeyFactory.stringToKey(userKey));
 
       DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
       PreparedQuery results = datastore.prepare(folderQuery);
 
       // Add folders to a list
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/UserFoldersServlet.java:50]: Comment not necessary.

File: src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 3 above)
Comment (tylerhou) [reply src/main/java/com/google/sps/servlets/LoginServlet.java:46]: Fixed in #5

File: src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 4 above)
Comment (tylerhou) [reply src/main/java/com/google/sps/servlets/LoginServlet.java:44]: Fixed in #5

File: src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 6 above)
Comment (tylerhou) [reply src/main/java/com/google/sps/servlets/LoginServlet.java:76]: Fixed in #5

File: src/test/java/com/google/sps/LoginServletTest.java (hunk 2 above)
Comment (tylerhou) [reply src/test/java/com/google/sps/LoginServletTest.java:90]: Fixed in #5

File: src/main/java/com/google/sps/data/Card.java (hunk 7 above)
Comment (ethyus) [reply src/main/java/com/google/sps/data/Card.java:104]: Yeah my bad. I think there was a miscommunication between the podmates that day. I didn't know Setup was the main branch on that day, so I ended up pushing some code. Currently the pushed code that you and Zach haven't gotten the chance to review is Folder.java, Card.java, User.java. However, I have edited them in other pull request, which I realized isn't good. I'm trying to minimize editing a file out of its respective PR.

File: src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 5 above)
Comment (ethyus) [reply src/main/java/com/google/sps/servlets/LoginServlet.java:57]: Should be fixed in #5 under "Restructure the json output in LoginServlet"

File: src/main/java/com/google/sps/data/Folder.java (hunk 9 above)
Comment (ethyus) [reply src/main/java/com/google/sps/data/Folder.java:55]: Fixed in #5  with recent commit

File: src/main/java/com/google/sps/servlets/UserFoldersServlet.java (hunk 11 above)
Comment (ethyus) [reply src/main/java/com/google/sps/servlets/UserFoldersServlet.java:50]: Fixed in #5 in recent commit

File: src/main/java/com/google/sps/servlets/UserCardsServlet.java (hunk 10 above)
Comment (ethyus) [reply src/main/java/com/google/sps/servlets/UserCardsServlet.java:84]: Fixed in #10 - will update this change in the design doc

File: src/main/java/com/google/sps/data/Card.java (hunk 8 above)
Comment (ethyus) [reply src/main/java/com/google/sps/data/Card.java:106]: Fixed in #10
